sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -86,6 +88,56 @@ case class Mode(
     buffer
   }
 
+  private def getCollationAwareBuffer(
+      childDataType: DataType,
+      buffer: OpenHashMap[AnyRef, Long]): Iterable[(AnyRef, Long)] = {
+    def groupAndReduceBuffer(groupingFunction: AnyRef => _): Iterable[(AnyRef, Long)] = {
+      buffer.groupMapReduce(t =>
+        groupingFunction(t._1))(x => x)((x, y) => (x._1, x._2 + y._2)).values
+    }
+    def determineBufferingFunction(
+        childDataType: DataType): Option[AnyRef => _] = {
+      childDataType match {
+        case _ if UnsafeRowUtils.isBinaryStable(child.dataType) => None
+        case _ => Some(collationAwareTransform(_, childDataType))
+      }
+    }
+    determineBufferingFunction(childDataType).map(groupAndReduceBuffer).getOrElse(buffer)
+  }
+
+  private def collationAwareTransform(myData: AnyRef, myElementType: DataType): AnyRef = {
+    myElementType match {
+      case _ if UnsafeRowUtils.isBinaryStable(myElementType) => myData
+      case st: StructType =>
+        processStructTypeWithBuffer(
+          myData.asInstanceOf[InternalRow].toSeq(
+            st).zip(
+            st.fields))
```

Comment:
this formatting looks a bit awkward, could you please correct it as per: https://github.com/apache/spark/pull/47154#discussion_r1724962435

Reply:
done!